src/raftstore/coprocessor/region_info_accessor.rs

```diff
@@ -328,7 +257,93 @@ impl RegionCollector {
         }
     }
 
+    /// Determines whether `region_to_check`'s epoch is stale compared to `current`'s epoch
+    #[inline]
+    fn is_region_epoch_stale(&self, region_to_check: &Region, current: &Region) -> bool {
+        let epoch = region_to_check.get_region_epoch();
+        let current_epoch = current.get_region_epoch();
+
+        epoch.get_version() < current_epoch.get_version()
+            || epoch.get_conf_ver() < current_epoch.get_conf_ver()
+    }
+
+    /// For all regions whose range overlaps with given `regions` or region_id is the same as
+    /// `region`'s, checks whether the given `region`'s epoch is not older than theirs.
+    /// Returns false if given `region` is stale, which means, at least one region above has newer
+    /// epoch. Returns true otherwise.
+    fn is_region_info_stale(&self, region: &Region) -> bool {
+        if let Some(region_with_same_id) = self.regions.get(&region.get_id()) {
+            if self.is_region_epoch_stale(region, &region_with_same_id.region) {
+                return true;
+            }
+        }
+
+        let regions_in_range = self
+            .region_ranges
+            .range((Excluded(data_key(region.get_start_key())), Unbounded));
+        for (_, id) in regions_in_range {
+            if *id == region.get_id() {
+                continue;
+            }
+
+            let current_region = &self.regions[id].region;
+            if !region.get_end_key().is_empty()
+                && current_region.get_start_key() >= region.get_end_key()
+            {
+                // This and following regions are not overlapping with `region`.
+                break;
+            }
+
+            if self.is_region_epoch_stale(region, current_region) {
+                return true;
+            }
+            // They are impossible to equal, or they cannot overlap.
```

Comment:
If there is duplicate message, will break this check?

Reply:
If two region has the same version and their range overlaps, then they must be the same region, if I didn't misunderstand anything... However here their region_ids must not equal, or the loop will be skipped at L286:
```
if *id == region.get_id() {
    continue;
}
```